File: format_excel.py
import openpyxl
import re

file_path = r"D:\Flutter_Project\DTCproduct\docs\2 Máy nén khí\CAC_LOI_THUONG_GAP_CUA_MNK.xlsx"
wb = openpyxl.load_workbook(file_path)

def format_text(text):
    if not text:
        return text
    text = str(text).strip()
    if not text:
        return text
    parts = re.split(r'[/_\n]', text)
    formatted = []
    for p in parts:
        p = p.strip()
        if not p:
            continue
        if p.startswith('-'): p = p[1:].strip()
        if p.startswith('+'): p = p[1:].strip()
        if p.startswith('_'): p = p[1:].strip()
        
        p = re.sub(r'[;:,.\s]+$', '', p)
        if p:
            p = p[0].upper() + p[1:]
            formatted.append(f'- {p}.')
    return '\n'.join(formatted)

def format_title(text):
    if not text:
        return text
    text = str(text).strip()
    text = re.sub(r'[;:,.\s]+$', '', text)
    if text:
        text = text[0].upper() + text[1:]
    return text

# Sheet 1: Lỗi_MNK_Thường_gặp
if 'Lỗi_MNK_Thường_gặp' in wb.sheetnames:
    ws = wb['Lỗi_MNK_Thường_gặp']
    for row in range(4, ws.max_row + 1):
        if ws.cell(row=row, column=2).value:
            ws.cell(row=row, column=2).value = format_title(ws.cell(row=row, column=2).value)
        if ws.cell(row=row, column=3).value:
            ws.cell(row=row, column=3).value = format_text(ws.cell(row=row, column=3).value)

# Sheet 2: Lỗi_MNK_HDSD_BĐK
if 'Lỗi_MNK_HDSD_BĐK' in wb.sheetnames:
    ws = wb['Lỗi_MNK_HDSD_BĐK']
    for row in range(4, ws.max_row + 1):
        if ws.cell(row=row, column=2).value:
            ws.cell(row=row, column=2).value = format_title(ws.cell(row=row, column=2).value)
        if ws.cell(row=row, column=3).value:
            ws.cell(row=row, column=3).value = format_text(ws.cell(row=row, column=3).value)
        if ws.cell(row=row, column=4).value:
            ws.cell(row=row, column=4).value = format_text(ws.cell(row=row, column=4).value)

# Sheet 3: Lỗi_MNK_Hao_Dầu
if 'Lỗi_MNK_Hao_Dầu' in wb.sheetnames:
    ws = wb['Lỗi_MNK_Hao_Dầu']
    for row in range(2, ws.max_row + 1):
        if ws.cell(row=row, column=2).value:
            ws.cell(row=row, column=2).value = format_title(ws.cell(row=row, column=2).value)
        if ws.cell(row=row, column=3).value:
            ws.cell(row=row, column=3).value = format_text(ws.cell(row=row, column=3).value)
        if ws.cell(row=row, column=4).value:
            ws.cell(row=row, column=4).value = format_text(ws.cell(row=row, column=4).value)

wb.save(file_path)
print("Excel updated")

File: format_excel.dart
import 'dart:io';
import 'package:excel/excel.dart';

String formatText(String text) {
  if (text.isEmpty) return '';
  var parts = text.split(RegExp(r'[/_\n]'));
  List<String> formatted = [];
  
  for (var p in parts) {
    p = p.trim();
    if (p.isEmpty) continue;
    
    if (p.startsWith('-')) p = p.substring(1).trim();
    if (p.startsWith('+')) p = p.substring(1).trim();
    if (p.startsWith('_')) p = p.substring(1).trim();
    
    p = p.replaceAll(RegExp(r'[;:,.\s]+$'), '');
    
    if (p.isNotEmpty) {
      p = p[0].toUpperCase() + p.substring(1);
      formatted.add('- $p.');
    }
  }
  return formatted.join('\n');
}

String formatTitle(String text) {
  if (text.isEmpty) return '';
  text = text.trim();
  text = text.replaceAll(RegExp(r'[;:,.\s]+$'), '');
  if (text.isNotEmpty) {
    text = text[0].toUpperCase() + text.substring(1);
  }
  return text;
}

void main() {
  var file = r"D:\Flutter_Project\DTCproduct\docs\2 Máy nén khí\CAC_LOI_THUONG_GAP_CUA_MNK.xlsx";
  var bytes = File(file).readAsBytesSync();
  var excel = Excel.decodeBytes(bytes);

  var sheet1 = excel.tables['Lỗi_MNK_Thường_gặp'];
  if (sheet1 != null) {
    for (int r = 3; r < sheet1.maxRows; r++) {
      var row = sheet1.row(r);
      if (row.length > 1 && row[1]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 1, rowIndex: r);
        sheet1.updateCell(cellIndex, TextCellValue(formatTitle(row[1]!.value.toString())));
      }
      if (row.length > 2 && row[2]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 2, rowIndex: r);
        sheet1.updateCell(cellIndex, TextCellValue(formatText(row[2]!.value.toString())));
      }
    }
  }

  var sheet2 = excel.tables['Lỗi_MNK_HDSD_BĐK'];
  if (sheet2 != null) {
    for (int r = 3; r < sheet2.maxRows; r++) {
      var row = sheet2.row(r);
      if (row.length > 1 && row[1]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 1, rowIndex: r);
        sheet2.updateCell(cellIndex, TextCellValue(formatTitle(row[1]!.value.toString())));
      }
      if (row.length > 2 && row[2]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 2, rowIndex: r);
        sheet2.updateCell(cellIndex, TextCellValue(formatText(row[2]!.value.toString())));
      }
      if (row.length > 3 && row[3]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 3, rowIndex: r);
        sheet2.updateCell(cellIndex, TextCellValue(formatText(row[3]!.value.toString())));
      }
    }
  }

  var sheet3 = excel.tables['Lỗi_MNK_Hao_Dầu'];
  if (sheet3 != null) {
    for (int r = 1; r < sheet3.maxRows; r++) {
      var row = sheet3.row(r);
      if (row.length > 1 && row[1]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 1, rowIndex: r);
        sheet3.updateCell(cellIndex, TextCellValue(formatTitle(row[1]!.value.toString())));
      }
      if (row.length > 2 && row[2]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 2, rowIndex: r);
        sheet3.updateCell(cellIndex, TextCellValue(formatText(row[2]!.value.toString())));
      }
      if (row.length > 3 && row[3]?.value != null) {
        var cellIndex = CellIndex.indexByColumnRow(columnIndex: 3, rowIndex: r);
        sheet3.updateCell(cellIndex, TextCellValue(formatText(row[3]!.value.toString())));
      }
    }
  }

  var newBytes = excel.encode();
  if (newBytes != null) {
    File(file).writeAsBytesSync(newBytes);
    print("Excel updated successfully");
  } else {
    print("Failed to encode excel");
  }
}
